src/omv/ports/mimxrt/omv_spi.c

```diff
@@ -40,36 +39,66 @@ typedef struct omv_spi_descr {
 static const omv_spi_descr_t omv_spi_descr_all[] = {
     #if defined(LPSPI1_ID)
     { LPSPI1, LPSPI1_SSEL_PIN,
-      { DMA0, DMAMUX, LPSPI1_DMA_TX_CHANNEL, kDmaRequestMuxLPSPI1Tx },
-      { DMA0, DMAMUX, LPSPI1_DMA_RX_CHANNEL, kDmaRequestMuxLPSPI1Rx } },
+      { DMA0, DMAMUX, kDmaRequestMuxLPSPI1Tx },
+      { DMA0, DMAMUX, kDmaRequestMuxLPSPI1Rx } },
     #else
-    { NULL, NULL, { NULL, NULL, 0, 0 }, { NULL, NULL, 0, 0 } },
+    { NULL, NULL, { NULL, NULL, 0 }, { NULL, NULL, 0 } },
     #endif
+
     #if defined(LPSPI2_ID)
     { LPSPI2, LPSPI2_SSEL_PIN,
-      { DMA0, DMAMUX, LPSPI2_DMA_TX_CHANNEL, kDmaRequestMuxLPSPI2Tx },
-      { DMA0, DMAMUX, LPSPI2_DMA_RX_CHANNEL, kDmaRequestMuxLPSPI2Rx } },
+      { DMA0, DMAMUX, kDmaRequestMuxLPSPI2Tx },
+      { DMA0, DMAMUX, kDmaRequestMuxLPSPI2Rx } },
     #else
-    { NULL, NULL, { NULL, NULL, 0, 0 }, { NULL, NULL, 0, 0 } },
+    { NULL, NULL, { NULL, NULL, 0 }, { NULL, NULL, 0 } },
     #endif
 
     #if defined(LPSPI3_ID)
     { LPSPI3, LPSPI3_SSEL_PIN,
-      { DMA0, DMAMUX, LPSPI3_DMA_TX_CHANNEL, kDmaRequestMuxLPSPI3Tx },
-      { DMA0, DMAMUX, LPSPI3_DMA_RX_CHANNEL, kDmaRequestMuxLPSPI3Rx } },
+      { DMA0, DMAMUX, kDmaRequestMuxLPSPI3Tx },
+      { DMA0, DMAMUX, kDmaRequestMuxLPSPI3Rx } },
     #else
-    { NULL, NULL, { NULL, NULL, 0, 0 }, { NULL, NULL, 0, 0 } },
+    { NULL, NULL, { NULL, NULL, 0 }, { NULL, NULL, 0 } },
     #endif
 
     #if defined(LPSPI4_ID)
     { LPSPI4, LPSPI4_SSEL_PIN,
-      { DMA0, DMAMUX, LPSPI4_DMA_TX_CHANNEL, kDmaRequestMuxLPSPI4Tx },
-      { DMA0, DMAMUX, LPSPI4_DMA_RX_CHANNEL, kDmaRequestMuxLPSPI4Rx } },
+      { DMA0, DMAMUX, kDmaRequestMuxLPSPI4Tx },
+      { DMA0, DMAMUX, kDmaRequestMuxLPSPI4Rx } },
     #else
-    { NULL, NULL, { NULL, NULL, 0, 0 }, { NULL, NULL, 0, 0 } },
+    { NULL, NULL, { NULL, NULL, 0 }, { NULL, NULL, 0 } },
     #endif
 };
 
+// LPSPI_MasterTransferEDMA doesn't have any code at all to generate a callback on transmit complete. So, we
+// add one here that simulates that. However, the DMA interrupt will occur before the SPI bus is done sending
+// data. So, we cannot call spi_master_callback below as it will happen before the data is actually sent. Instead,
+// we use the SPI transfer complete interrupt to call spi_master_callback at the right time. Note that we have
+// to zero out the descriptor to prevent LPSPI_MasterTransferHandleIRQ from doing anything we don't want it to do.
```

Comment:
I'm not following, it seems it does have code to call back on transmit complete and this is it (transferDone will be set to true), but why do we need this ?

Reply:
They only implemented a callback for RX. There's literally no code in the driver for a TX callback. As for how you are supposed to know when the TX transfer completes... no clue. I read the driver code and it was clearly missing though.

Reply:
It seems to be called after every transfer (tx or rx), and when it's done transferDone should be set. But I'm still not sure why we need this callback ?

Reply:
As mentioned, they don't setup the callback for TX in their code. That's why I have to do it manually. They only setup the callback for RX.

Reply:
E.g. I have to do this to enable the TX callback.

```
 if (xfer->txbuf) {
            // There isn't a race condition here to worry about since the major interrupt status flag
            // will remain asserted even if the DMA transfer was to complete before we enable the interrupt.
            EDMA_SetCallback(&spi->dma_descr_tx, EDMA_LpspiMasterTxCallback, spi);
            EDMA_EnableChannelInterrupts(spi->dma_descr_tx.base, spi->dma_descr_tx.channel,
                                         (uint32_t) kEDMA_MajorInterruptEnable);
        }
```

Reply:
We seem to be going in circles here... What is the purpose of this callback ? I know it's a TX complete callback, but why do we care when TX is complete ? The callback seems to zero the struct after the TX is done, why is that needed ?

Reply:
See the large 4 line comment above the function.

I have to zero out the descriptor to prevent unwanted behavior from occurring in the driver HAL. Then enable the SPI transfer complete interrupt so they happen in the right order.

The DMA interrupt happens before the SPI bus finishes sending data. So, I can't do anything from it. However, I can enable the transfer complete interrupt for the SPI. Which when it fires will run callback code.

If I don't zero most of the descriptor values then it will have side affects and not just call my the callback. See `void LPSPI_MasterTransferHandleIRQ(LPSPI_Type *base, lpspi_master_handle_t *handle)` in the fsl_lpspi.c file.

Reply:
Note, you can't use the transfer complete interrupt directly from the SPI since that fires per frame sent. You have to get the transfer complete from the DMA first and then enable the interrupt for the SPI to known the end of the data transfer to the physical wire.

Reply:
Updated comment as requested.